xdem/coreg.py

```diff
@@ -1442,3 +1482,457 @@ def _to_matrix_func(self) -> np.ndarray:
             raise NotImplementedError
         else:
             raise ValueError("A 2nd degree or higher ZScaleCorr cannot be described as a 4x4 matrix!")
+
+class PiecewiseCoreg(Coreg):
+    """
+    Piece-wise coreg class for nonlinear estimations.
+
+    A coreg class of choice is run on an arbitrary subdivision of the raster. When later applying the coregistration,\
+        the optimal warping is interpolated based on X/Y/Z shifts from the coreg algorithm at the grid points.
+
+    E.g. a subdivision of 4 means to divide the DEM in four equally sized parts. These parts are then coregistered\
+        separately, creating four Coreg.fit results. If the subdivision is not divisible by the raster shape,\
+        subdivision is made as best as possible to have approximately equal pixel counts.
+    """
+
+    def __init__(self, coreg: Coreg | CoregPipeline, subdivision: int, success_threshold: float = 0.8, n_threads: int | None = None):
+        """
+        Instantiate a piecewise coreg object.
+
+        :param coreg: An instantiated coreg object to fit in the subdivided DEMs.
+        :param subdivision: The number of chunks to divide the DEMs in. E.g. 4 means four different transforms.
+        :param success_threshold: Raise an error if fewer chunks than the fraction failed for any reason.
+        :param n_threads: The maximum amount of threads to use. Default=auto
+        """
+        if isinstance(coreg, type):
+            raise ValueError(
+                    "The 'coreg' argument must be an instantiated Coreg subclass. "
+                    "Hint: write e.g. ICP() instead of ICP"
+            )
+        self.coreg = coreg
+        self.subdivision = subdivision
+        self.success_threshold = success_threshold
+        self.n_threads = n_threads
+
+        super().__init__()
+
+        self._meta["coreg_meta"] = []
+
+
+
+    def _fit_func(self, ref_dem: np.ndarray, tba_dem: np.ndarray, transform: rio.transform.Affine | None,
+                  weights: np.ndarray | None, verbose: bool = False):
+        """Fit the coreg approach for each subdivision."""
+
+        groups = self.subdivide_array(tba_dem.shape)
+
+        indices = np.unique(groups)
+
+        progress_bar = tqdm(total=indices.size, desc="Coregistering chunks", disable=(not verbose))
+
+        def coregister(i: int) -> dict[str, Any] | BaseException:
+            """Coregister a chunk in a thread-safe way."""
+            inlier_mask = groups == i
+
+            # Find the corresponding slice of the inlier_mask to subset the data
```

Comment:
Looks somewhat similar (although more succinct?) than the function I added in spatial_tools: https://github.com/GlacioHack/xdem/blob/c60330abf5847be3654411374ae5bdf2a3d151d2/xdem/spatial_tools.py#L58

Reply:
Ah yes, there it is! I looked in geoutils and didn't find it so I gave up. But that's exactly what I needed.